### PR TITLE
Implement rule Avoid Link Headers

### PR DESCRIPTION
**Pull request by ishishov**

### CODE REVIEW

File: src/main/java/de/zalando/zally/rules/AvoidLinkHeadersRule.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
+package de.zalando.zally.rules;
+
+import de.zalando.zally.Violation;
+import de.zalando.zally.ViolationType;
+import io.swagger.models.*;
```
Comment (netme) [inline src/main/java/de/zalando/zally/rules/AvoidLinkHeadersRule.java]: Can we avoid `*` imports?

File: src/main/java/de/zalando/zally/rules/AvoidLinkHeadersRule.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+    @Override
+    public List<Violation> validate(Swagger swagger) {
+        List<Violation> violations = new ArrayList();
+        if (swagger != null && swagger.getPaths() != null) {
+            for (Map.Entry<String, Path> path : swagger.getPaths().entrySet())
+                for (Map.Entry<HttpMethod, Operation> op : path.getValue().getOperationMap().entrySet()) {
+                    for (Parameter param : op.getValue().getParameters()) {
+                        if ("header".equals(param.getIn()) && "Link".equals(param.getName())) {
+                            String violationPath = path.getKey() + "/" + op.getKey() + "/" + param.getName();
```
Comment (netme) [inline src/main/java/de/zalando/zally/rules/AvoidLinkHeadersRule.java]: Is it possible to somehow simplify this loop? Currently we have 5 levels, maybe it's good to move some logic into some subfunctions?

File: src/test/java/de/zalando/zally/rules/AvoidLinkHeadersRuleTest.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        AvoidLinkHeadersRule rule = new AvoidLinkHeadersRule();
+        Swagger swagger = new SwaggerParser().read("/api_spa.yaml");
+        assertThat(rule.validate(swagger)).isEmpty();
+    }
+
+    @Test
+    public void avoidLinkHeadersInvalidJson() throws IOException {
+        AvoidLinkHeadersRule rule = new AvoidLinkHeadersRule();
+        Swagger swagger = new SwaggerParser().read("/fixtures/avoidLinkHeaderRuleInvalid.json");
+        assertThat(rule.validate(swagger).size()).isEqualTo(2);
```
Comment (netme) [inline src/test/java/de/zalando/zally/rules/AvoidLinkHeadersRuleTest.java]: Maybe it's also good to test which exact violations are happening

File: src/main/java/de/zalando/zally/rules/AvoidLinkHeadersRule.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+                            violations.add(new Violation(TITLE, DESCRIPTION, VIOLATION_TYPE, RULE_LINK, violationPath));
+                        }
+                    }
+                    for (Map.Entry<String, Response> response : op.getValue().getResponses().entrySet()) {
+                        if (response.getValue().getHeaders() != null && response.getValue().getHeaders().containsKey("Link")) {
+                            String violationPath = path.getKey() + "/" + op.getKey() + "/" + response.getKey();
```
Comment (netme) [inline src/main/java/de/zalando/zally/rules/AvoidLinkHeadersRule.java]: You can also use `String.join()` here

File: src/test/java/de/zalando/zally/rules/AvoidLinkHeadersRuleTest.java (hunk 3 above)
Comment (mkulak) [inline src/test/java/de/zalando/zally/rules/AvoidLinkHeadersRuleTest.java]: I suggest to add validation for Violation's fields also.

File: src/main/java/de/zalando/zally/rules/AvoidLinkHeadersRule.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+public class AvoidLinkHeadersRule implements Rule {
+    private static final String TITLE = "Avoid Link in Header Rule";
+    private static final String DESCRIPTION = "Do Not Use Link Headers with JSON entities";
+    private static final ViolationType VIOLATION_TYPE = ViolationType.MUST;
+    private static final String RULE_LINK = "http://zalando.github.io/restful-api-guidelines/hyper-media/Hypermedia.html#must-do-not-use-link-headers-with-json-entities";
+
+    @Override
+    public List<Violation> validate(Swagger swagger) {
+        List<Violation> violations = new ArrayList();
+        if (swagger != null && swagger.getPaths() != null) {
```
Comment (superbenji) [inline src/main/java/de/zalando/zally/rules/AvoidLinkHeadersRule.java]: can we include the check for null for the swagger object in the controller as it effects all rules?

File: src/main/java/de/zalando/zally/rules/AvoidLinkHeadersRule.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+                    for (Parameter param : op.getValue().getParameters()) {
+                        if ("header".equals(param.getIn()) && "Link".equals(param.getName())) {
+                            String violationPath = path.getKey() + "/" + op.getKey() + "/" + param.getName();
+                            violations.add(new Violation(TITLE, DESCRIPTION, VIOLATION_TYPE, RULE_LINK, violationPath));
+                        }
+                    }
+                    for (Map.Entry<String, Response> response : op.getValue().getResponses().entrySet()) {
+                        if (response.getValue().getHeaders() != null && response.getValue().getHeaders().containsKey("Link")) {
+                            String violationPath = path.getKey() + "/" + op.getKey() + "/" + response.getKey();
+                            violations.add(new Violation(TITLE, DESCRIPTION, VIOLATION_TYPE, RULE_LINK, violationPath));
```
Comment (mkulak) [inline src/main/java/de/zalando/zally/rules/AvoidLinkHeadersRule.java]: Please refactor common code among 25-27 and 31-33 in separate method.

File: src/main/java/de/zalando/zally/rules/AvoidLinkHeadersRule.java (hunk 1 above)
Comment (ishishov) [inline src/main/java/de/zalando/zally/rules/AvoidLinkHeadersRule.java]: looks like IDE automatically does it when there's 5+ imports from the same package, reverted

File: src/main/java/de/zalando/zally/rules/AvoidLinkHeadersRule.java (hunk 2 above)
Comment (ishishov) [inline src/main/java/de/zalando/zally/rules/AvoidLinkHeadersRule.java]: there are 3 loops levels + conditions which is ok, 'for' loops in this case is more readable than lambda on streams. But I will move violation creation in a separate method

File: src/main/java/de/zalando/zally/rules/AvoidLinkHeadersRule.java (hunk 5 above)
Comment (ishishov) [inline src/main/java/de/zalando/zally/rules/AvoidLinkHeadersRule.java]: I think we can add it to the controller, but it doesn't mean it shouldn't be here as well, right?

File: src/test/java/de/zalando/zally/rules/AvoidLinkHeadersRuleTest.java (hunk 3 above)
Comment (ishishov) [inline src/test/java/de/zalando/zally/rules/AvoidLinkHeadersRuleTest.java]: will compare the validation name and type also

**Comment by netme** [comment]
Would be great if you'll squash latest commits. 

```
git rebase -i 	6146614
````

**Pull request merged**